### ISSUE TITLE
prevent direct calls to sdag entry functions

### ISSUE THREAD

**Issue by jcphill** [UIUC-PPL/charm#738]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/738*

---
Calls to sdag entries from within "serial" blocks of sdag entries return immediately (or at the first unfulfilled "when" block), like a message send, rather than integrating into the sdag workflow.  This is not the expected behavior.  Giving the function that implements an sdag entry method a different name than the entry method itself would prevent this, so it would be impossible to accidentally call mysdag() rather than thisProxy->mysdag().

This issue was reported by one of the speakers at the Charm++ workshop.

**Comment by PhilMiller** [comment]
*Original date: 2015-05-21 18:30:38*

---
This is a serious issue of ease of use and principle of least surprise that we need to work out. We'll discuss it, and try to find something to do about it.

**Comment by ericjbohm** [comment]
*Original date: 2015-06-25 19:33:00*

---
Figure out how to handle the case where the user really wants this functionality vs the case where someone stumbles into unexpectedly by nesting.

**Comment by PhilMiller** [comment]
*Original date: 2015-06-25 19:35:26*

---
Per core meeting discussion, this is a semantic issue that needs to be worked out carefully, not just a coding task. Meeting between Phil, Jonathan, & Eric M to work out what to do about this.

**Comment by nitbhat** [comment]
*Original date: 2017-02-06 20:59:36*

---
Modified every sdag entry method to be implemented in .C as "_sdag_" + entry method name

In a .ci file if there's an entry method called mysdag() that is used with the when construct, it is implemented with a function "_sdag_mysdag". This prevents direct calls made to "mysdag()". 

Gerrit:  ~~https://charm.cs.illinois.edu/gerrit/#/c/2205/~~ https://github.com/UIUC-PPL/charm/commit/7c127d8c5473a4aee6622cc0c167731570a4a5a5

**Comment by nitbhat** [comment]
*Original date: 2017-06-05 18:44:12*

---
Changed the prefix for sdag entry methods to "_sdag_fnc_" as "_sdag_init" is an already generated xi function.

**Comment by PhilMiller** [comment]
*Original date: 2017-06-13 20:49:50*

---
Proposed API change, lots of corresponding code to change, not a priority for getting the release out.

**Comment by lvkale** [comment]
*Original date: 2017-11-09 16:16:53*

---
Phil's note 2 years ago said "is a semantic issue that needs to be worked out carefully" and suggested a meeting. Is this a correct, consensus decision? (A minor issue is projections will show unexpected EP name. But that is easy to handle).

**Comment by nitbhat** [comment]
*Original date: 2017-11-09 17:22:01*

---
Fix: Gerrit: ~~https://charm.cs.illinois.edu/gerrit/#/c/2205/~~ https://github.com/UIUC-PPL/charm/commit/7c127d8c5473a4aee6622cc0c167731570a4a5a5
The current implemented fix is where an sdag entry method that contains when statement(s) gets renamed to "_sdag_fnc_"+name and thereby cannot be directly called. Sdag entry methods without when statements can still be directly called.  

We did discuss it in core a few months back, but it'll be worthwhile to bring it up again esp because OpenAtom uses direct sdag calls and those will have to change. 

Openatom build failure: (http://ppl-jenkins:8080/job/test-openatom/9872/console)

**Comment by lvkale** [comment]
*Original date: 2017-11-09 17:41:20*

---
Ideally, charmxi should also produce an error message (in case the function name fname doesn’t match, but _sdag_fnc_fname matches) telling the programmer not to call sdag. Is that easy/hard/too hard?

**Comment by nitbhat** [comment]
*Original date: 2017-11-09 18:41:32*

---
Didn't entirely understand the example. Do you mean when the user declares an SDAG Entry method in the .ci as "fname" and then calls the xi generated method "_sdag_fnc_fname" directly? (instead of calling proxy.fname();)

**Comment by nitbhat** [comment]
*Original date: 2017-12-08 18:17:34*

---
This is waiting on the changes to be made on Openatom. (~~https://charm.cs.illinois.edu/gerrit/#/c/2205/~~ https://github.com/UIUC-PPL/charm/commit/7c127d8c5473a4aee6622cc0c167731570a4a5a5)

**Comment by nitbhat** [comment]
*Original date: 2018-03-20 20:57:36*

---
I have made changes to Openatom to convert direct sdag calls to proxy calls and have this change on a branch called `nitin/sdag_changes` in the Openatom repository. Need an openatom expert to review and merge this.

**Comment by nitbhat** [comment]
*Original date: 2018-04-02 19:53:36*

---
Openatom changes corresponding to this bug fix have been merged: https://charm.cs.illinois.edu/gerrit/#/c/openatom/+/3917/
